### ISSUE TITLE
[no-unnecessary-boolean-literal-compare] False positive with nullable types

### ISSUE THREAD

**Issue by lpa-vadim-zaslavsky** [typescript-eslint/typescript-eslint#1632]
**Repro**
```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error"
  }
}
```
```TS
let index: { [key: string]: boolean | undefined };
const isTrue = (key: string) => index[key] === true;
```
**Expected Result**
The rule is not triggered as the type is `boolean | undefined`. See https://github.com/palantir/tslint/issues/4602.

**Actual Result**
> This expression unnecessarily compares a boolean value to a boolean instead of using it directly

**Additional info**
TSConfig flags:
```JSON
{
    "target": "es5",
    "module": "esnext",
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "downlevelIteration": true,
    "noEmitHelpers": true,
    "importHelpers": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "strictFunctionTypes": true,
    "preserveSymlinks": true
}
```

**Versions**
| package                                        | version |
| ------------------------------------ | -------- |
| `@typescript-eslint/eslint-plugin` | `2.20.0` |
| `@typescript-eslint/parser`           | `2.20.0` |
| `TypeScript`                                   | `3.7.5`   |
| `ESLint`                                          | `6.8.0`   |
| `node`                                           | `13.8.0` |
| `npm`                                            | `6.13.7` |

**Comment by bradzacher** [comment]
You are not using `strictNullChecks`.
This means that typescript essentially erases the `| undefined` from your type.

I cannot repro this against master because we test with the flag on.
You'll have to turn it on as well if you want this to work.

**Comment by lpa-vadim-zaslavsky** [comment]
Thanks for the response.

I'm not sure about erasing `undefined` from the type definition, but after having a look at the [docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#--strictnullchecks) on `strictNullChecks` I can see that the rule cannot work without the flag because

> `T` and `T | undefined` are considered synonymous in regular type checking mode (because `undefined` is considered a subtype of any `T`)
